### ISSUE TITLE
Specialization

### ISSUE THREAD

**Issue by WilsonGramer** [wipplelang/wipple#95]
Allow using the `[specialize]` attribute on a generic constant to give it an alternate implementation for a more specific type.

## Syntax

For example, the `count` function can be made more performant for lists who know their size without needing to iterate over their elements:

```wipple 
count :: Collection where (Iterate Collection _) => Collection -> Natural
count : ...

[specialize]
count :: List _ -> Natural
count : ...

list (1 , 2 , 3) . count -- uses the `specialize`d constant instead of the original
```

The specialized constant must have the same name as the original constant, and its type must unify with the original constant's type (including bounds).

**Comment by WilsonGramer** [comment]
Partially implemented in ae6c66e — the typechecker doesn't yet check that the specialized constant unifies with the original constant.

**Comment by WilsonGramer** [comment]
Additional checks implemented in 23813c0.
